File: src/puppy_bringup/scripts/odom_tf_publisher.py
#!/usr/bin/env python3
"""
Puppy 机械狗 - 里程计 TF 发布器
订阅 /odom 话题，发布 odom -> base_footprint 变换
（CoppeliaSim 的 simROS2 不支持 tf2_msgs/msg/TFMessage，所以用此节点转发）
注意：simROS2 使用 best_effort QoS，需要匹配
"""
import rclpy
from rclpy.node import Node
from rclpy.qos import QoSProfile, ReliabilityPolicy, HistoryPolicy
from nav_msgs.msg import Odometry
from geometry_msgs.msg import TransformStamped
from tf2_ros import TransformBroadcaster


class OdomTfPublisher(Node):
    def __init__(self):
        super().__init__('odom_tf_publisher')
        self.tf_broadcaster = TransformBroadcaster(self)
        # simROS2 使用 best_effort QoS，需要匹配
        qos = QoSProfile(
            reliability=ReliabilityPolicy.BEST_EFFORT,
            history=HistoryPolicy.KEEP_LAST,
            depth=10
        )
        self.odom_sub = self.create_subscription(
            Odometry, '/odom', self.odom_callback, qos)
        self.get_logger().info('里程计 TF 发布器已启动，订阅 /odom (best_effort QoS)')

    def odom_callback(self, msg):
        t = TransformStamped()
        t.header.stamp = msg.header.stamp
        t.header.frame_id = msg.header.frame_id  # odom
        t.child_frame_id = msg.child_frame_id    # base_footprint
        t.transform.translation.x = msg.pose.pose.position.x
        t.transform.translation.y = msg.pose.pose.position.y
        t.transform.translation.z = msg.pose.pose.position.z
        t.transform.rotation = msg.pose.pose.orientation
        self.tf_broadcaster.sendTransform(t)


def main(args=None):
    rclpy.init(args=args)
    node = OdomTfPublisher()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()


if __name__ == '__main__':
    main()

File: cpp_src/puppy_bringup/odom_tf_publisher.cpp
// 里程计 TF 发布器实现
#include "puppy_bringup/odom_tf_publisher.h"

namespace puppy_bringup {

OdomTfPublisherNode::OdomTfPublisherNode()
    : rclcpp::Node("odom_tf_publisher") {
  tf_broadcaster_ = std::make_shared<tf2_ros::TransformBroadcaster>(this);

  // simROS2 使用 best_effort QoS，需要匹配
  rclcpp::QoS qos(10);
  qos.best_effort();
  qos.keep_last(10);

  odom_sub_ = this->create_subscription<nav_msgs::msg::Odometry>(
      "/odom", qos,
      std::bind(&OdomTfPublisherNode::onOdom, this, std::placeholders::_1));

  RCLCPP_INFO(this->get_logger(),
              "里程计 TF 发布器已启动，订阅 /odom (best_effort QoS)");
}

void OdomTfPublisherNode::onOdom(
    const nav_msgs::msg::Odometry::SharedPtr msg) {
  geometry_msgs::msg::TransformStamped t;
  t.header.stamp = msg->header.stamp;
  t.header.frame_id = msg->header.frame_id;    // odom
  t.child_frame_id = msg->child_frame_id;      // base_footprint
  t.transform.translation.x = msg->pose.pose.position.x;
  t.transform.translation.y = msg->pose.pose.position.y;
  t.transform.translation.z = msg->pose.pose.position.z;
  t.transform.rotation = msg->pose.pose.orientation;
  tf_broadcaster_->sendTransform(t);
}

}  // namespace puppy_bringup
